Apps/Src/ReadCarCAN.c

```diff
@@ -5,256 +5,482 @@
+#define ENABLE_SATURATION_MSG 1
 
 // CAN watchdog timer variable
 static OS_TMR canWatchTimer;
 
-// prechargeDlyTimer
-static OS_TMR prechargeDlyTimer;
+// Precharge delay timer variable
+static OS_TMR arrayBypassPrechargeDlyTimer;
 
-// Array restart thread lock
-static OS_MUTEX arrayRestartMutex;
-static bool restartingArray = false;
+// Precharge delay timer variable
+static OS_TMR motorControllerBypassPrechargeDlyTimer;
 
 // NOTE: This should not be written to anywhere other than ReadCarCAN. If the need arises, a mutex to protect it must be added.
 // Indicates whether or not regenerative braking / charging is enabled.
-static bool chargeEnable = false;
+static bool chargeEnable = false; // Enable (High message) of BPS high voltage (HV) array contactor
+
+// BPS HV Array saturation buffer variables
+static int8_t HVArrayChargeMsgBuffer[SAT_BUF_LENGTH];
+static int8_t HVArrayMsgSaturation = 0;
+static uint8_t HVArrayOldestMsgIdx = 0;
+static int8_t HVPlusMinusChargeMsgBuffer[SAT_BUF_LENGTH];
+static int8_t HVPlusMinusChargeMsgSaturation = 0;
+static uint8_t HVPlusMinusMotorOldestMsgIdx = 0;
 
-// Saturation buffer variables
-static int8_t chargeMsgBuffer[SAT_BUF_LENGTH];
-static int chargeMsgSaturation = 0;
-static uint8_t oldestMsgIdx = 0;
+// Array ignition (IGN_1) and Motor Controller ignition (IGN_2) pin status
+static bool arrayIgnitionStatus = false;
+static bool motorControllerIgnitionStatus = false;
 
-// SOC and Supp V
+// Boolean to indicate precharge status for Array Precharge Bypass Contactor (PBC) and Motor Controller PBC
+static bool arrayBypassPrechargeComplete = false;
+static bool motorControllerBypassPrechargeComplete = false;
+
+// State of Charge (SOC) and supplemental battery pack voltage (SBPV) value intialization
 static uint8_t SOC = 0;
 static uint32_t SBPV = 0;
 
+// Error assertion function prototype
+static void assertReadCarCANError(ReadCarCAN_error_code_t rcc_err);
 
-// Handler to turn array back on
-static void arrayRestart(void *p_tmr, void *p_arg); 
+// Getter function for charge enable, indicating that battery charging is allowed
+bool ChargeEnable_Get(void){
+    return chargeEnable;
+}
 
+/**
+ * @brief Nested function as the same function needs to be cexecuted however the timer requires different parameters
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+*/
+static void callbackCANWatchdog(void *p_tmr, void *p_arg){
+   assertReadCarCANError(READCARCAN_ERR_MISSED_MSG);
+}
 
-// Getter function for chargeEnable
-bool ChargeEnable_Get() 
-{
-    return chargeEnable;
+/**
+ * @brief Callback function for the precharge delay timer. Waits for precharge and then sets arrayBypassPrechargeComplete to true.
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+*/
+static void setArrayBypassPrechargeComplete(void *p_tmr, void *p_arg){
```

Comment:
We discussed this, and the reason we cannot have a OSTmrStateGet is because the timer and the array/mc PBC complete boolean do not reflect each other. This is because array/mc PBC complete bool is set to true when precharge timer is complete, but unlike precharge timer, the bool is then set to false when array/mc contactor is on. When precharge happens again, the array/mc PBC complete boolean is set to ON again, if conditions are met.